go/store/blobstore/oci.go

```diff
@@ -0,0 +1,796 @@
+// For InMemoryBlobstore instances error should never be returned (though other
+// implementations of this interface can)
+func (bs *OCIBlobstore) Exists(ctx context.Context, key string) (bool, error) {
+	absKey := path.Join(bs.prefix, key)
+	namespace, err := bs.getNamespace(ctx)
+	if err != nil {
+		return false, err
+	}
+	_, err = bs.client.HeadObject(ctx, objectstorage.HeadObjectRequest{
+		NamespaceName: &namespace,
+		BucketName:    &bs.bucketName,
+		ObjectName:    &absKey,
+	})
+	if err == nil {
+		return true, nil
+	}
+	return false, nil
+}
+
+// Get retrieves an io.reader for the portion of a blob specified by br along with its version
+func (bs *OCIBlobstore) Get(ctx context.Context, key string, br BlobRange) (io.ReadCloser, string, error) {
+	absKey := path.Join(bs.prefix, key)
+	namespace, err := bs.getNamespace(ctx)
+	if err != nil {
+		return nil, "", err
+	}
+
+	req := objectstorage.HeadObjectRequest{
+		NamespaceName: &namespace,
+		BucketName:    &bs.bucketName,
+		ObjectName:    &absKey,
+	}
+
+	res, err := bs.client.HeadObject(ctx, req)
+	if err != nil {
+		if serr, ok := common.IsServiceError(err); ok {
+			// handle not found code
+			if serr.GetHTTPStatusCode() == 404 {
+				return nil, "", NotFound{"oci://" + path.Join(bs.bucketName, absKey)}
+			}
+		}
+		return nil, "", err
+	}
+
+	requestedSize, totalSize, offSet, err := getDownloadInfo(br, res.ContentLength)
+	if err != nil {
+		return nil, "", NotFound{"oci://" + path.Join(bs.bucketName, absKey)}
+	}
+
+	etag := fmtstr(res.ETag)
+
+	return bs.multipartDownload(ctx, absKey, namespace, etag, requestedSize, totalSize, offSet)
+}
+
+func (bs *OCIBlobstore) multipartDownload(ctx context.Context, absKey, namespace, etag string, requestedSize, totalSize, offSet int64) (io.ReadCloser, string, error) {
+	parts, err := downloadParts(ctx, absKey, namespace, etag, requestedSize, totalSize, defaultBatchSize, offSet, defaultPartSize, maxPartNum, defaultConcurrentListeners, bs.downloadPart)
+	if err != nil {
+		return nil, "", err
+	}
+	rc, err := assembleParts(parts)
+	return rc, etag, err
+}
+
+func (bs *OCIBlobstore) getObject(ctx context.Context, absKey, namespace, etag string, br BlobRange) (io.ReadCloser, string, error) {
+	req := objectstorage.GetObjectRequest{
+		NamespaceName: &namespace,
+		BucketName:    &bs.bucketName,
+		ObjectName:    &absKey,
+	}
+
+	byteRange := br.asHttpRangeHeader()
+	if byteRange != "" {
+		req.Range = &byteRange
+	}
+
+	if etag != "" {
+		req.IfMatch = &etag
+	} else {
+		star := "*"
+		req.IfNoneMatch = &star
+	}
+
+	res, err := bs.client.GetObject(ctx, req)
+	if err != nil {
+		if serr, ok := common.IsServiceError(err); ok {
+			// handle not found code
+			if serr.GetHTTPStatusCode() == 404 {
+				return nil, "", NotFound{"oci://" + path.Join(bs.bucketName, absKey)}
+			}
+		}
+		return nil, "", err
+	}
+
+	// handle negative offset and positive length
+	if br.offset < 0 && br.length > 0 {
+		trimmedR, err := trimContent(res, br.length)
```

Comment:
Why isn't this just a LimitReader wrapper? Because we might not get the full response and we want to consider that an error?